src/common/cockpitjson.c

```diff
@@ -967,3 +967,165 @@ cockpit_json_to_hash_table (JsonObject *object,
+cockpit_json_walk_node (JsonNode    *node,
+                        WalkClosure *closure);
+
+static JsonArray *
+cockpit_json_walk_array (JsonArray   *array,
+                         WalkClosure *closure)
+{
+  gint n = json_array_get_length (array);
+  JsonArray *copy = NULL;
+
+  for (gint i = 0; i < n; i++)
+    {
+      JsonNode *element = json_array_get_element (array, i);
+      JsonNode *node = cockpit_json_walk_node (element, closure);
+
+      if (node && !copy)
+        {
+          /* We've decided we need to copy now, so revisit everything up
+           * to this point and add it to the copy.
+           */
+          copy = json_array_sized_new (n);
+          for (gint j = 0; j < i; j++)
+            json_array_add_element (copy, json_array_dup_element (array, j));
+        }
+
+      if (copy)
+        json_array_add_element (copy, node ?: json_node_ref (element));
+    }
+
+  return copy;
+}
+
+static JsonObject *
+cockpit_json_walk_object (JsonObject  *object,
+                          WalkClosure *closure)
+{
+  JsonObject *copy = NULL;
+  JsonObjectIter iter;
+  const gchar *key;
+  JsonNode *value;
+
+  json_object_iter_init (&iter, object);
+  while (json_object_iter_next (&iter, &key, &value))
+    {
+      JsonNode *node = cockpit_json_walk_node (value, closure);
+
+      if (node && !copy)
+        {
+          const gchar *this_key = key;
+
+          /* We've decided we need to copy now, so revisit everything up
+           * to this point and add it to the copy.
+           */
+          copy = json_object_new ();
+          json_object_iter_init (&iter, object);
+          while (json_object_iter_next (&iter, &key, &value) && key != this_key)
+            json_object_set_member (copy, key, json_node_ref (value));
+
+          /* ...now key and value will be where they were before we entered */
+          g_assert (key == this_key);
+        }
+
+      if (copy)
+        json_object_set_member (copy, key, node ?: json_node_ref (value));
+    }
+
+  return copy;
+}
+
+static JsonNode *
+cockpit_json_walk_node (JsonNode    *node,
+                        WalkClosure *closure)
+{
+  JsonNode *new_node = NULL;
+
+  switch (json_node_get_node_type (node))
+    {
+    case JSON_NODE_ARRAY:
+      {
+        JsonArray *array = json_node_get_array (node);
+        JsonArray *copy = cockpit_json_walk_array (array, closure);
+
+        if (copy)
+          {
+            new_node = json_node_new (JSON_NODE_ARRAY);
+            json_node_take_array (new_node, copy ?: json_array_ref (array));
+          }
+
+        break;
+      }
+
+    case JSON_NODE_OBJECT:
+      {
+        JsonObject *object = json_node_get_object (node);
+        JsonObject *copy = cockpit_json_walk_object (object, closure);
+
+        if (copy)
+          {
+            new_node = json_node_new (JSON_NODE_OBJECT);
+            json_node_take_object (new_node, copy ?: json_object_ref (object));
+          }
+
+        break;
+      }
+
+    default:
```

Comment:
Please make this `JSON_NODE_VALUE`; explicitly handle `JSON_NODE_NULL`, and make the  `default:` a `g_assert_not_reached()`. Please add a unit test for `null` values.

Reply:
It was my intend that `JSON_NODE_NULL` also be handed to the callback.  To be honest, if I could come up with a way to also handle the container cases with the callback, then I would... the limitation that we only call for value types is ... not great.

If we did do a change here, I would actually indeed try to expand the callback to be called for every type, with the following semantics for containers:

 - `NULL` return: do the iteration
 - non-`NULL` return: take this value in place
 - the callback might call back into `_walk()` on the subtree in order to do some combination of the above

Reply:
Right, I agree that the callback should also be called on `NULL`, I would just like this code to be explicit about it, and not do "anything else". It is arguably unlikely that there ever will be more JSON_NODE_* types, but I still find it easier to read if the cases are explicit.

Reply:
I'm ok with that `default:` cleanup to happen in a follow-up. It's not actually changing functionality at this point.